index.helper.ts

```diff
@@ -0,0 +1,306 @@
+            };
+            if ('start' in e && 'end' in e) {
+              Object.assign(res, {
+                start: e.start,
+                end: e.end,
+              });
+            }
+            return res;
+          });
+
+        const formattedSlots = this.formatSlots(s.text, sampleEntities);
+
+        // Add language as a property at the same level as intent
+        return {
+          text: s.text,
+          intent: valueMap[intent.value].value,
+          language: s.language.code, // Add language here
+          slots: formattedSlots,
+        };
+      });
+
+    return dataset;
+  }
+
+  private formatSlots(
+    text: string,
+    entities: LudwigNlu.ExampleEntity[],
+  ): string {
+    const words = text.split(/\s+/);
+
+    // Initialize the slots array with 'O' tags
+    const slots = Array(words.length).fill('O');
+
+    // Track the current character position in the original text
+    let currentPosition = 0;
+
+    // Iterate over the words and map them to slots using entity indices
+    words.forEach((word, index) => {
+      // Calculate the start and end indices of the current word
+      const wordStart = currentPosition;
+      const wordEnd = currentPosition + word.length;
+
+      // Look for a matching entity whose indices overlap the current word
+      const matchingEntity = entities.find(
+        (e) => e.start < wordEnd && e.end > wordStart, // Check for overlap
+      );
+
+      if (matchingEntity) {
+        slots[index] = `B-${matchingEntity.entity}`;
+      }
+
+      // Update the current position (account for the space after the word)
+      currentPosition = wordEnd + 1;
+    });
+
+    const formattedSlots = slots.join(' ');
+    return formattedSlots;
+  }
+
+  /**
+   * Perform a training request
+   *
+   * @param samples - Samples to train
+   * @param entities - All available entities
+   * @returns The training result
+   */
+  async train(
+    samples: NlpSampleFull[],
+    entities: NlpEntityFull[],
+  ): Promise<any> {
+    throw new Error('Method not Implemented yet');
+  }
+
+  /**
+   * Perform evaluation request
+   *
+   * @param samples - Samples to evaluate
+   * @param entities - All available entities
+   * @returns Evaluation results
+   */
+  async evaluate(
+    samples: NlpSampleFull[],
+    entities: NlpEntityFull[],
+  ): Promise<any> {
+    throw new Error('Method Not Implemented Yet');
+  }
+
+  async processIncomingNluPayload(
+    nlp: LudwigNlu.LudwigNluResultType,
+    givenText: string,
+  ): Promise<LudwigNlu.NluProcessedResultType> {
+    const words = givenText.split(' ');
+
+    const intentValue = nlp.intent.predictions.intent_predictions;
+    const intentKey = `intent_probabilities_${intentValue}`; // e.g., "intent_probabilities_greeting"
+    const intentConfidence = nlp.intent.predictions[intentKey];
+    const formattedIntentPayload = {
+      name: intentValue,
+      confidence: intentConfidence,
+    };
```

Comment:
.We need to parse the results depending on the trait entities we have (we may not only have intent)